### ISSUE TITLE
Compare microkernel benchmark IR against main for PRs

### ISSUE THREAD

**Issue by alexbaden** [intel/intel-xpu-backend-for-triton#2974]
It would be useful to be able to detect whether or not the generated IR (Triton IR or TTGIR) changes for a given pull request (code change). For example, if the IR changes for a particular microkernel then we should have high suspicion for performance changes for that microkernel. Performance measurements are often noisy, whereas IR changes are essentially binary. This would also be useful for when we sync openai commits or make changes to the pass pipeline and want to know all the places that could be affected. 

The proposal here is to implement a CI job that runs on every PR and compiles the Triton IR for the microkernel benchmark to TTIR/TTGIR using `triton.compile`. To keep things simple and keep runtime down, we will use the nightly wheels for our "golden reference". A suggested flow is:

0. Create script which runs the microkernel benchmark IR. Perhaps we can import this from the microkernel benchmarks directly? 
1. Get latest Triton nightly wheels and run the microkernel benchmark IR script to generate TTIR/TTGIR for each microkernel. Note that a given microkernel benchmark may actually generate multiple TTIR/TTGIR. We can start with 1 though and improve as needed.
2. Run the same script as part of the build_and_test pipeline on the rolling driver. 
3. Diff the IRs. 

Note that many PRs will be expected to change the IR, so we do not want the job to fail if the IR differs (otherwise we will have a lot of red PRs that are actually fine, and merging might be blocked). But, it would be nice to have some way of notifying the user when the IR differs instead of having to remember to check the output of a given github actions run.

**Comment by whitneywhtsang** [comment]
The motivation of this issue is good, we want to have a way to quickly identify if performance impact is variance or not.

> Note that a given microkernel benchmark may actually generate multiple TTIR/TTGIR. We can start with 1 though and improve as needed.

What's your expected behavior when there are more than 1 TTIR/TTGIR?

In general, I worry that this approach may be too fragile, and end up reporting differences in most cases, and then developers start to ignore the report. MLIR changes all the time, if we use Triton nightly wheels output as baseline, there can be many PRs merged between that and the PR, or the PR could be based on a commit before the Triton nightly wheels.

**Comment by whitneywhtsang** [comment]
> Eventually I would like to be able to diff as much IR as possible, but to start I think grabbing one representative set of IRs from each microbenchmark would be ok. I am not sure how to best grab the IR without running the benchmark, figuring that out is part of this issue.

Just to clarify, do you mean diffing the final MLIR of one input shape?

> Do you have another approach that would be less noisy that we could try?

If we reconsider the motivation, which is to have a way to quickly identify if performance impact is variance or not. Another way to resolve that could be CI automatically determine if the performance results are within standard deviation, and report only gains and degradations. Or CI automatically rerun benchmarks with potential performance impact.

**Comment by alexbaden** [comment]
The motivation is to be able to determine if a given PR is going to change the TTIR/TTGIR (and possibly even the MLIR). Detecting performance regressions, etc is just one benefit from knowing if a change is mutating the IR. 

Yes, it seems like most microbenchmarks do parameter sweeps across input shapes but I imagine there are other possible parametrizations.

**Comment by sommerlukas** [comment]
Offline, I mentioned some scripts from the LLVM infrastructure that can help to abstract over SSA value names. 
* For MLIR, there is this script: https://github.com/llvm/llvm-project/blob/main/mlir/utils/generate-test-checks.py
* For LLVM IR, there is: https://github.com/llvm/llvm-project/blob/main/llvm/utils/update_test_checks.py

Even if we don't use the scripts themselves, we can maybe learn something from how they detect and abstract SSA value names.

**Comment by arunjose696** [comment]
The plan for now is to just have a on option to report if IR changed or not and if IR changed have a link which shows the diff in a meaningfull way which will be figured out by checking the common differences. For now we would not be raising issues or marking PRs as red, this will be done later based on how the differences look like.

**Comment by arunjose696** [comment]
To compare IRs effectively, a reliable ground truth is necessary. Currently, nightly builds are stored as artifacts from GitHub Actions runs. However, each nightly build wheel is approximately 1.4 GB, making it impractical to build and store wheels for every commit. Using just nightly builds as the ground truth would also flag every commit made on a day when IR changes occur as different, leading to numerous false positives and reducing reliability.

Two alternative approaches to consider are:

Storing IRs for Each Commit:
IRs can be stored as artifacts of GitHub Actions runs. For example, running gemm_benchmark.py generates over 200 IR folders in the cache, with a total size of 18 MB. This could be further optimized by limiting the benchmark suite to a single set of parameters, potentially reducing storage requirements to just a few MB per commit.

Generating IRs from Previous Commits During CI:
The GitHub Actions environment contains the full repository, including the Git history. This allows checking out previous commits and running benchmarks to generate IRs for comparison. While this approach avoids storing additional files, it increases runtime due to the need to install Triton and rerun benchmarks.

Any suggestions on which of these 2 choose?

**Comment by alexbaden** [comment]
I don't think we want to store IRs - being able to compare results from two different commits either by using pre-built binaries is a useful feature in and of itself. For running Triton I think we can shrink the size of the stored build file dramatically. 

But for now let's focus on splitting a task out of this issue to come up with a way to compare two previously generated IRs. We can work on how we are going to generate the IRs in parallel. I think we should get a feeling for what the diffs look like for a few different sets of IRs over a few different time periods so we can decide how best to proceed.
